zgw-consumers refactor

SonnyBA opened the pull request:
Fixes #7 
Partially fixes https://github.com/maykinmedia/open-api-framework/issues/66

Upgrading zgw-consumers to ensure we can upgrade zgw consumers in Open Zaak: https://github.com/open-zaak/open-zaak/pull/1800

stevenbal committed:
:alien: Update NotificationsConfig client usage

notifications-api-common has changed its underlying client to ape_pie.APIClient due to a zgw-consumers update (https://github.com/maykinmedia/notifications-api-common/pull/15)

SonnyBA committed:
Update project requirements

To make them match with notifications-api-common>=0.3.0

SonnyBA committed:
Add ape-pie & zgw-consumers requirements

SonnyBA committed:
Use  to do external API calls

SonnyBA committed:
Remove APICredentials model

SonnyBA committed:
Remove zds_client usage

SonnyBA committed:
Run formatting

SonnyBA committed:
Update tox configuration

SonnyBA committed:
Update documentation

SonnyBA committed:
Add reverse migration path

SonnyBA committed:
Update docs

SonnyBA committed:
Update CI configuration

stevenbal commented on vng_api_common/client.py:
```diff
@@ -1,17 +1,67 @@
+    return response_json
+
+
+class Client(APIClient):
```
If possible, I would prefer if we use `zgw_consumers.client.ZGWClient` for this, I think it does everything we need: https://github.com/maykinmedia/zgw-consumers/blob/ce82492a8d82703345f057c24359827b69cc1a6a/zgw_consumers/client.py#L91-L114

SonnyBA replied:
The link points to the `ZGWAuth` class but the `ZGWClient` was [moved](https://github.com/maykinmedia/zgw-consumers/commit/48c8f8ed1cc6e406886a9d775ea3694665df3f28#diff-0607b90d5b0cd2063b12f5c914bb297d1810c90610b1366aac9fff4765d8f029L34) to the `legacy` module and will be [removed](https://zgw-consumers.readthedocs.io/en/latest/changelog.html#id9) in the future.

stevenbal replied:
ah you're right, I thought I removed this comment, I was mainly referring to the auth_headers part, but I added another comment for that

stevenbal commented on vng_api_common/client.py:
```diff
@@ -1,17 +1,67 @@
+        return super().request(method, url, *args, **kwargs)
+
+
+def get_client(url: str) -> Client | Any | None:
```
Do we need this custom `get_client` with the `CLIENT_CLASS` setting? I think we should be able to use `build_client` from zgw_consumers together with the `ZGWAuth` class to create a client

Open Forms does this as well: https://github.com/open-formulieren/open-forms/blob/683efe79d912e977da79a6bddff5d077ceee28fd/src/openforms/registrations/contrib/zgw_apis/client.py#L23-L26

SonnyBA replied:
We could remove the setting and use the `build_client` function. This will prevent applications to use their own client of course but that might be a considered decision

SonnyBA replied:
Answered through https://github.com/maykinmedia/commonground-api-common/pull/38#discussion_r1816459670

stevenbal commented on vng_api_common/client.py:
```diff
@@ -25,20 +75,38 @@ def get_client(url: str, url_is_api_root=False) -> Optional[Client]:
+    return build_client(service.api_root, client_factory=client_class)
+
+
+def get_auth_headers(
```
I think we should be able to make use of https://github.com/maykinmedia/zgw-consumers/blob/main/zgw_consumers/client.py#L91

stevenbal commented on vng_api_common/oas.py:
```diff
@@ -1,8 +1,5 @@
 """
```
I'm not 100% sure, but we might be able to remove this file, since the schema fetching is no longer used?

stevenbal replied:
Decided to keep it, as we still use some of the utils in openzaak tests

annashamray replied:
`obj_has_shape` is even used in loose-fk validators in Open Zaak, but i'm not sure about schema fetcher

stevenbal commented on vng_api_common/models.py:
```diff
@@ -71,86 +69,6 @@ def __str__(self):
-        return auth
-
-
 class ClientConfig(SingletonModel):
```
I think we should be able to get rid of this, it's essentially a really stripped down version of Service, so it would be nice if we can rely on that instead

stevenbal replied:
I think (but it might be good to double check this) that the `NotificationsConfig` from `commonground-api-common` is no longer used, the config has been moved to `notifications_api_common` and there we do use `Service`: https://github.com/maykinmedia/notifications-api-common/blob/main/notifications_api_common/models.py#L16

I think we should use the same pattern for `AuthorizationsConfig`

stevenbal commented on vng_api_common/models.py:
```diff
@@ -166,9 +84,9 @@ def save(self, *args, **kwargs):
         """
         config = cls.get_solo()
-        return _get_client(config.api_root, url_is_api_root=True)
+        return _get_client(config.api_root)
```
This now fails when checking the config for autorisaties API for instance, because `build_client` expects a `Service`, not an `api_root`:

```
File "/home/steven/Documents/commonground-api-common/vng_api_common/views.py", line 104, in get_context_data
    config += _test_ac_config()
              ^^^^^^^^^^^^^^^^^
  File "/home/steven/Documents/commonground-api-common/vng_api_common/views.py", line 132, in _test_ac_config
    ac_client = AuthorizationsConfig.get_client()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/steven/Documents/commonground-api-common/vng_api_common/models.py", line 92, in get_client
    return _get_client(config.api_root)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/steven/Documents/commonground-api-common/vng_api_common/client.py", line 89, in get_client
    return build_client(service.api_root, client_factory=client_class)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/steven/Documents/VNG/open-notificaties/env/lib/python3.11/site-packages/zgw_consumers/client.py", line 30, in build_client
    config_adapter, nlx_base_url=service.nlx, **kwargs
                                 ^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'nlx'
```

stevenbal commented on vng_api_common/notifications/models.py:
```diff
@@ -91,7 +85,18 @@ def register(self) -> None:
+        except ClientError as exc:
+            response_json = exc.args[0]
+
+            raise RuntimeError(
```
Not sure what errors were possible previously, but I don't think `RuntimeError` was one of them? I think I'd just log and raise the `RequestException`

stevenbal commented on vng_api_common/utils.py:
```diff
@@ -191,9 +191,14 @@ def request_object_attribute(
 ) -> str:
     client = get_client(url)
 
+    if not client:
+        return ""
```
I haven't seen this function `request_object_attribute` before, it might be worth checking if this is actually used in any projects (or other libraries)

SonnyBA committed:
Describe `APICredentials` to `Service` model migration

SonnyBA commented on setup.cfg:
```diff
@@ -34,22 +33,23 @@ scripts =
-    drf-yasg>=1.20.0
-    drf-nested-routers>=0.93.3
-    gemma-zds-client>=0.14.0
+    drf-yasg>=1.20.0 # TODO: remove usage?
```
Looks like https://github.com/maykinmedia/commonground-api-common/issues/29 was made for this

SonnyBA committed:
Add missing  field for Service's

SonnyBA commented on vng_api_common/models.py:
```diff
@@ -166,9 +84,9 @@ def save(self, *args, **kwargs):
 
     @classmethod
-    def get_client(cls) -> Optional[Client]:
+    def get_client(cls) -> APIClient | Any | None:
```
Whenever you set `CLIENT_CLASS` anything other than the default, it will not return the default (`APIClient` at this time of writing)

SonnyBA replied:
I guess the underlying question is if we want users to allow them to use clients that do not inherit from `APIClient`. If we want to allow users to do this, `Any` would be correct here, otherwise it should not be typed like that.

stevenbal replied:
I don't think we'll be using clients that do not inherit from `APIClient`, so let's use `APIClient | None`

SonnyBA commented on vng_api_common/client.py:
```diff
@@ -1,17 +1,67 @@
+    except RequestException as exc:
+        if response.status_code >= 500:
+            raise
+        raise ClientError(response_json if response_json is not None else {}) from exc
```
We could but we might lose some interesting information whenever requests are not done correctly (e.g `400` error status codes from responses). This could be used for logging purposes like it was done before, see [7fa83c5dc0e294fe430c819f9993532cc878293f](https://github.com/maykinmedia/gemma-zds-client/commit/7fa83c5dc0e294fe430c819f9993532cc878293f) (if an application does not want to use django-log-outgoing-requests and use its own).

SonnyBA committed:
Try to determine `api_type` from `api_root`

This can be useful for Openzaak instances where these paths are used

SonnyBA committed:
Reuse `client.auth` to retrieve token

SonnyBA committed:
Remove oas related validators

SonnyBA committed:
Fix `build_client` usage

SonnyBA committed:
Remove `request_object_attribute`

SonnyBA committed:
Fix import error

SonnyBA committed:
Remove `notifications-api-common` related definitions

These definitions are included in `notifications-api-common`

SonnyBA committed:
Remove notifications-api-common related models & update
AuthorizationsConfig model

Removed models should be imported from notifications-api-common

SonnyBA committed:
Fix migrations, use simple `get_client` function, migrate
AuthorizationsConfig

SonnyBA committed:
Update JWTAuth error handling

SonnyBA committed:
Fix import errors

stevenbal committed:
Revert "Remove oas related validators"

This reverts commit 726b28a7ad2528ef8557c86213ace0979c3f0d2d.

stevenbal committed:
:memo: Re-add OAS utils docs

stevenbal committed:
:bug: Fix publishvalidator by adding missing imports

the imports and the publishvalidator tests were removed

stevenbal committed:
:recycle: Replace get_auth_headers with generate_jwt util

which can be reused in Open Zaak/Open Notificaties

stevenbal committed:
:white_check_mark: Add migration tests for service migrations

stevenbal commented on vng_api_common/migrations/0006_delete_apicredential.py:
```diff
@@ -0,0 +1,106 @@
+    return APITypes.orc
+
+
+def migrate_credentials_to_service(apps, _) -> None:
```
@annashamray I'm not sure if these APICredentials were actually still used, so I'm not sure if this migration is actually needed tbh

annashamray replied:
This migrations looks very similar to `autorisaties.0003_fill_service` migration. I think they were already migrated
I don't think APICredentials are used in Open Zaak. but there is a high change that they are still used in Open Notificaties

annashamray replied:
oops sorry I thought it was Open Zaak repo. Please ignore my comment about duplicated migration. 

I've checked and Open Notificaties indeed uses APICredential, for example to access Autorisaties API

SonnyBA reviewed (commented):
Github won't let me approve (because I'm the owner of the PR), but this looks good :+1:

annashamray commented on vng_api_common/models.py:
```diff
@@ -69,106 +61,3 @@ class Meta:
         return self.identifier
-
-
-class APICredential(models.Model):
```
There should be some warning in the CHANGELOG in the future, that the model to change credentials has been changed.

annashamray commented on vng_api_common/models.py:
```diff
@@ -69,106 +61,3 @@ class Meta:
-        return auth
-
-
-class ClientConfig(SingletonModel):
```
Why is this abstract model deleted? I'm pretty sure it's used in Open Zaak. Maybe just change api_root to be a FK to `Service` model instead of the url

annashamray commented on vng_api_common/notifications/constants.py:
```diff
@@ -1,3 +0,0 @@
-# Exernally defined scopes.
-SCOPE_NOTIFICATIES_CONSUMEREN_LABEL = "notificaties.consumeren"
```
I understand that these labels are duplicated now in `notifications-api-common`, but it looks like cleaning them is out of scope of this PR
Maybe a separate PR or at least a warning in CHANGELOG would be useful, because it could really be a breaking change

stevenbal replied:
I checked Open Zaak and Open Notificaties, the former already imports them from `notifications-api-common` and the latter doesn't use them. I'll add a warning in the changelog nonetheless

annashamray commented on vng_api_common/notifications/models.py:
```diff
@@ -1,97 +0,0 @@
-import uuid
```
Same comment as for notification labels.

stevenbal replied:
added it to the warning in changelog as well

annashamray commented on vng_api_common/validators.py:
```diff
@@ -240,15 +240,22 @@ def __call__(self, informatieobject: str, serializer):
-        except requests.HTTPError as exc:
+
+            oios = to_internal_data(response)
+        except (requests.RequestException, RuntimeError) as exc:
```
Why don't you catch ClientError here? Or is it caught with `requests.RequestException` ?

stevenbal replied:
It seems that `ClientError` is based on `RuntimeError`, but I'll change it to explicitly catch `ClientError` instead :+1:

stevenbal replied:
Upon checking, this validator isn't used anywhere in Open Zaak (which uses its own `CreateRemoteRelationValidator`), so I'll remove it

annashamray commented on vng_api_common/utils.py:
```diff
@@ -12,10 +12,10 @@
-from zds_client.client import ClientError
 
-from .client import get_client
+from .client import get_client, to_internal_data
```
These imports are never used

annashamray commented on vng_api_common/authorizations/models.py:
```diff
@@ -1,27 +1,62 @@
+        ),
+        verbose_name=_("autorisations api service"),
+        on_delete=models.SET_NULL,
+        blank=True,
+        null=True,
```
I do think it should not be nullable.

stevenbal replied:
I've added a migration to remove null=True

SonnyBA committed:
Add zgw-consumers-oas to requirements

Used for the ResourceValidator, which verifies if objects have the
correct shape as defined in an OAS.

SonnyBA committed:
call `raise_for_status` when testing AC & NRC connections

SonnyBA commented:
@stevenbal I added https://github.com/maykinmedia/commonground-api-common/pull/38/commits/44a40320f9a6f3506bbd1f7a9d2a2c117a528fb5 today, as I noticed the configuration view was broken (incorrectly stated services were reachable) by the changes made in this PR. They were not picked up by the tests though, so I also added/updated some tests.

stevenbal committed:
:ok_hand: [maykinmedia/open-api-framework#66] PR feedback

stevenbal committed:
:heavy_plus_sign: Add missing requests-mock dependency

SonnyBA committed:
Handle duplicate service slugs

SonnyBA committed:
Apply formatting

annashamray reviewed (approved):
Looks good!

annashamray commented on setup.cfg:
```diff
@@ -34,22 +33,25 @@ scripts =
-    notifications-api-common>=0.2.2
+    notifications-api-common>=0.3.0
+    zgw-consumers>=0.35.1
+    zgw-consumers-oas
```
Why do we need this new dependency? Now we will use client which doesn't request OAS right?

SonnyBA replied:
@stevenbal I changed this requirement in an [separate PR](https://github.com/maykinmedia/commonground-api-common/pull/40/commits/b420445442379a51c5aa35bd5de70d2c8393115d) (which points to this branch)

stevenbal replied:
this mean that projects that make use of these OAS utils have to install `commonground-api-common[tests]`

SonnyBA committed:
Move `zgw-consumers-oas` to test requirements

annashamray commented on vng_api_common/views.py:
```diff
@@ -172,8 +181,7 @@ def _test_nrc_config() -> list:
     nrc_config = NotificationsConfig.get_solo()
-
-    nrc_client = NotificationsConfig.get_client()
+    nrc_client: Optional[Client] = NotificationsConfig.get_client()
```
I think on the line 178 `notifications_api_common` app should be checked

stevenbal replied:
good catch! I'll change it

stevenbal committed:
Merge pull request #40 from maykinmedia/slug-migrations

Handle duplicate service slugs

stevenbal committed:
:bug: Check for correct installed notifs app in nrc config check

this checked for vng_api_common.notifications, but should check notifications_api_common since the notifications code moved there
